src/main/java/cloud/orbit/actors/extensions/metrics/dropwizard/OrbitMessagingMetricsExtension.java

```diff
@@ -74,46 +74,52 @@ public OrbitMessagingMetricsExtension(final String name, final String beforeHand
-        sendMeters.put((int) MessageDefinitions.RESPONSE_ERROR, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:response_error,direction:outbound]"));
-        sendMeters.put((int) MessageDefinitions.RESPONSE_OK, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:response_ok,direction:outbound]"));
-        sendMeters.put((int) MessageDefinitions.RESPONSE_PROTOCOL_ERROR, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:response_protocol_error,direction:outbound]"));
+        inboundMetrics.put((int) MessageDefinitions.ONE_WAY_MESSAGE, MetricsManager.getInstance().getRegistry().timer("orbit.messaging[type:one_way_message,direction:inbound]"));
+        inboundMetrics.put((int) MessageDefinitions.REQUEST_MESSAGE, MetricsManager.getInstance().getRegistry().timer("orbit.messaging[type:request_message,direction:inbound]"));
+        inboundMetrics.put((int) MessageDefinitions.RESPONSE_ERROR, MetricsManager.getInstance().getRegistry().timer("orbit.messaging[type:response_error,direction:inbound]"));
+        inboundMetrics.put((int) MessageDefinitions.RESPONSE_OK, MetricsManager.getInstance().getRegistry().timer("orbit.messaging[type:response_ok,direction:inbound]"));
+        inboundMetrics.put((int) MessageDefinitions.RESPONSE_PROTOCOL_ERROR, MetricsManager.getInstance().getRegistry().timer("orbit.messaging[type:response_protocol_error,direction:inbound]"));
+
+        outboundMetrics.put((int) MessageDefinitions.ONE_WAY_MESSAGE, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:one_way_message,direction:outbound]"));
+        outboundMetrics.put((int) MessageDefinitions.REQUEST_MESSAGE, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:request_message,direction:outbound]"));
+        outboundMetrics.put((int) MessageDefinitions.RESPONSE_ERROR, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:response_error,direction:outbound]"));
+        outboundMetrics.put((int) MessageDefinitions.RESPONSE_OK, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:response_ok,direction:outbound]"));
+        outboundMetrics.put((int) MessageDefinitions.RESPONSE_PROTOCOL_ERROR, MetricsManager.getInstance().getRegistry().meter("orbit.messaging[type:response_protocol_error,direction:outbound]"));
     }
 
     @Override
-    public void onRead(HandlerContext ctx, Object message)
+    public void onRead(HandlerContext ctx, Object object)
     {
-        if (message instanceof Message)
+        long now = System.currentTimeMillis();
+        if (object instanceof Message)
         {
-            Meter meter = receiveMeters.get(((Message) message).getMessageType());
-            if (null != meter)
+            Message message = (Message) object;
+            Long messageCreationTimestamp = (Long) message.getHeader(MESSAGING_METRICS_HEADER_TIMESTAMP);
```

Comment:
I question if this cast will work when a message has been deserialized over the wire. Would the header be a String and as such should be it a Long.valueOf()?